src/fromager/sdist.py

```diff
@@ -259,6 +278,150 @@ def handle_requirement(
     return resolved_version
 
 
+def _resolve_from_graph(
+    ctx: context.WorkContext,
+    prev_graph: DependencyGraph | None,
+    parent_req: Requirement | None,
+    req: Requirement,
+    req_type: RequirementType,
+    pre_built: bool,
+) -> tuple[str, Version] | None:
+    # we have already resolved top level reqs before bootstrapping
+    # so they should already be in the root node
+    if req_type == RequirementType.TOP_LEVEL:
+        for edge in ctx.dependency_graph.get_root_node().get_outgoing_edges(
+            req.name, RequirementType.TOP_LEVEL
+        ):
+            if edge.req == req:
+                return (
+                    edge.destination_node.download_url,
+                    edge.destination_node.version,
+                )
+        # this should never happen
+        return None
+
+    if not prev_graph:
+        return None
+
+    seen_version = set()
+
+    # first perform resolution using the top level reqs before looking at history
+    possible_versions_in_top_level = []
+    for top_level_edge in ctx.dependency_graph.get_root_node().get_outgoing_edges(
+        req.name, RequirementType.TOP_LEVEL
+    ):
+        possible_versions_in_top_level.append(
+            (
+                top_level_edge.destination_node.download_url,
+                top_level_edge.destination_node.version,
+            )
+        )
+        seen_version.add(str(top_level_edge.destination_node.version))
+
+    resolver_result = _resolver_for_graph(ctx, possible_versions_in_top_level, req)
```

Comment:
This function is named like it's going to return a resolver, but it's used like it's returning results.